src/Compilers/CSharp/Test/Semantic/Semantics/ExtensionTypeTests.cs

```diff
@@ -0,0 +1,4993 @@
+public class Container<T>
+{
+    public class C<U> where U : T
+    {
+        T M() => throw null;
+    }
+}
+class C2
+{
+    Container<int[]> M2() => throw null;
+    void M3(Container<int[]>.C<int[]> c, Container<int[]>.C<byte[]> c2, Container<int[]>.C<long[]> c3) { }
+}
+""";
+        var comp = CreateCompilation(src);
+        comp.VerifyDiagnostics(
+            // (11,69): error CS0311: The type 'byte[]' cannot be used as type parameter 'U' in the generic type or method 'Container<int[]>.C<U>'. There is no implicit reference conversion from 'byte[]' to 'int[]'.
+            //     void M3(Container<int[]>.C<int[]> c, Container<int[]>.C<byte[]> c2, Container<int[]>.C<long[]> c3) { }
+            Diagnostic(ErrorCode.ERR_GenericConstraintNotSatisfiedRefType, "c2").WithArguments("Container<int[]>.C<U>", "int[]", "U", "byte[]").WithLocation(11, 69),
+            // (11,100): error CS0311: The type 'long[]' cannot be used as type parameter 'U' in the generic type or method 'Container<int[]>.C<U>'. There is no implicit reference conversion from 'long[]' to 'int[]'.
+            //     void M3(Container<int[]>.C<int[]> c, Container<int[]>.C<byte[]> c2, Container<int[]>.C<long[]> c3) { }
+            Diagnostic(ErrorCode.ERR_GenericConstraintNotSatisfiedRefType, "c3").WithArguments("Container<int[]>.C<U>", "int[]", "U", "long[]").WithLocation(11, 100)
+            );
+        var containerR = comp.GetMember<MethodSymbol>("C2.M2").ReturnType;
+        var c = containerR.GetTypeMembers().Single();
+        Assert.Equal("Container<System.Int32[]>.C<U>", c.ToTestDisplayString());
+        var substitutedTypeParameter = c.TypeParameters.Single();
+        Assert.False(substitutedTypeParameter.IsArray());
+    }
+
+    [Fact]
+    public void Attributes()
+    {
+        var src = """
+class MyAttribute : System.Attribute { }
+class C { }
+
+[My]
+explicit extension R for C
+{
+    [My] void M() { }
+}
+""";
+        // PROTOTYPE what attribute target should be used for extensions?
+        var comp = CreateCompilation(src);
+        comp.VerifyDiagnostics(
+            // (4,2): error CS0592: Attribute 'My' is not valid on this declaration type. It is only valid on 'assembly, module, class, struct, enum, constructor, method, property, indexer, field, event, interface, parameter, delegate, return, type parameter' declarations.
+            // [My]
+            Diagnostic(ErrorCode.ERR_AttributeOnBadSymbolType, "My").WithArguments("My", "assembly, module, class, struct, enum, constructor, method, property, indexer, field, event, interface, parameter, delegate, return, type parameter").WithLocation(4, 2)
+            );
+
+        var r = comp.GlobalNamespace.GetTypeMember("R");
+        Assert.Equal("MyAttribute", r.GetAttributes().Single().ToString());
+
+        var m = r.GetMethod("M");
+        Assert.Equal("MyAttribute", m.GetAttributes().Single().ToString());
+    }
+
+    [Fact]
+    public void ReservedTypeNames()
+    {
```

Comment:
Consider testing with other restricted words, such as `file` and `required`. Also consider testing unconditional keywords.